doc/dev/backend-plugins.md

```diff
@@ -0,0 +1,271 @@
+If a backend plugin does not support caching, it should immediately return a value indicating that the cache is invalid.
+If there is no cache entry for this backend, `libelektra-kdb` skips this phase.
+
+During the `cachecheck` phase the backend plugin is called with:
+
+- The exact `parentKey` that was returned by the `resolver` phase of this `get` operation.
+  The key name and value of this key are read-only.
+  Additionally, the metakey `internal/kdb/cachetime` is set to a value indicating the update time of the cache entry.
+  TODO: exact format of `internal/kdb/cachetime` TBD
+- An empty keyset `ks`.
+
+The backend plugin then:
+
+- **MUST** indicate via the return value, whether the cache entry for this backend is still valid.
+
+> **Note**: The backend plugin may also modify the keyset `ks`, but `libelektra-kdb` will discard this keyset after this phase, so these modifications won't have any effects.
+
+#### Storage Phases
+
+These phases are responsible for reading and validating the actual data stored in the KDB.
+
+In the `prestorage` phase the backend plugin is called with:
+
+- The exact `parentKey` that was returned by the `resolver` phase of this `get` operation.
+  The key name and value of this key are read-only.
+- An empty keyset `ks`.
+
+There are no restrictions on what the backend plugin may do in this phase, but just like in the `resolver` phase, change to `ks` will be discarded.
+This phase is useful for file-level manipulations, like file-based encryption, line ending conversion or verifying file signatures.
+In this sense, it is the counter-part of the `precommit` phase of the `set` operation.
+
+The `storage` phase is the where the actual data is read.
+In this phase the backend plugin is called with:
+
+- The exact `parentKey` that was returned by the `prestorage` phase of this `get` operation.
+  The key name and value of this key are read-only.
+- An empty keyset `ks`.
+
+The backend plugin then:
+
+- **MUST** read the data for the mountpoint from the storage unit that was selected in the `resolver` phase.
+- **MUST** parse that data and insert it below `parentKey` into `ks`.
+
+The last of the storage phases is the `poststorage` phase.
+In this phase the backend plugin is called with:
+
+- The exact `parentKey` that was returned by the `storage` phase of this `get` operation.
+  The key name and value of this key are read-only.
+- The exact `ks` that was returned by the `storage` phase of this `get` operation.
+
+Again there are no restrictions on what the backend plugin may do in this phase.
+However, unlike the `prestorage` phase, this phase is a very important one.
+It is where validation, generation of implicit values and similar tasks happen.
+
+Finally, `libelektra-kdb` merges the keyset returned by the `poststorage` phase with the ones returned by other backend plugins for different mountpoints and then returns it to the user.
+
+TODO: Are modifications to `parentKey` visible to the user?
```

Comment:
`kdb file` requires that the file name remains in the parentKey.

Reply:
1. I tried to be agnostic to the type of backend. It doesn't have to be a file name.
2. A few lines above this I already defined, that both key name and key value are read-only after the `resolver` phase. So the key value (currently the file name) cannot be changed.
3. This was mostly about whether or not modifying the metadata of `parentKey` (other than `error` and `warning`) should be allowed. And if these changes would stay after the end of `kdbGet()`, or if `kdbGet()` should clear the metadata (except `error` and `warning`).

Reply:
Other modifications (besides the value in resolver) are currently not needed. But I agree that we should have defined the semantics when they are allowed. I think we should be rather strict, we can always loosen the requirements (if a use case actually turns up).

Reply:
> But I agree that we should have defined the semantics when they are allowed.

Exactly.

IMO metadata (other than `error`/`warning`) on `parentKey` should be discarded or not allowed at all. In theory there are some use-cases, but there are better solutions for those (e.g. global keyset).

Reply:
> metadata (other than error/warning) on parentKey should be discarded or not allowed at all. In theory there are some use-cases, but there are better solutions for those (e.g. global keyset).

I agree.